Fix to Issue #6

ashutoshshrm529 opened the pull request:
Added the ellipse, circle and elliptical arc functions and tested locally.

ashutoshshrm529 committed:
Ellipse Function Added. Need to test.

ashutoshshrm529 committed:
Fix issue #6. Added Ellipse, Circle and Elliptical Arc as mentioned.

ashutoshshrm529 commented:
Since there are no examples in this repository, the checks are not successful. Please Merge https://github.com/arduino-libraries/ArduinoGraphics/pull/10 or add examples before merging.

per1234 commented on keywords.txt:
```diff
@@ -33,6 +33,9 @@ line	KEYWORD2
 rect	KEYWORD2
+ellipse KEYWORD2
+circle KEYWORD2
+arc KEYWORD2
```
```suggestion
ellipse	KEYWORD2
circle	KEYWORD2
arc	KEYWORD2
```
The Arduino IDE will only recognize keywords when a true tab (not spaces) separator is used.

Reference:
https://arduino.github.io/arduino-cli/library-specification/#keywordstxt-format

per1234 commented on src/ArduinoGraphics.cpp:
```diff
@@ -184,6 +184,95 @@ void ArduinoGraphics::rect(int x, int y, int width, int height)
+  int x2 = x1 + r1 - 1;
+
+  for (x = x1; x <= x2; x++) {
+      y2 = (int) (y1 + sqrt((1 - ((x * x)/(r1 * r1)))) * r2);
```
```suggestion
      int y2 = (int) (y1 + sqrt((1 - ((x * x)/(r1 * r1)))) * r2);
```
Lack of variable declaration causes compilation to fail.

per1234 commented on src/ArduinoGraphics.cpp:
```diff
@@ -184,6 +184,95 @@ void ArduinoGraphics::rect(int x, int y, int width, int height)
+  int x2 = x1 + radius;
+
+  for (x = x1; x <= x2; x++) {
+      y2 = (int) y1 + sqrt((radius*radius)-(x*x)) ;
```
```suggestion
      int y2 = (int) y1 + sqrt((radius*radius)-(x*x)) ;
```

per1234 commented on src/ArduinoGraphics.cpp:
```diff
@@ -184,6 +184,95 @@ void ArduinoGraphics::rect(int x, int y, int width, int height)
+  }
+}
+
+void ArduinoGraphics::arc(int x, int y, int radiusX, int radiusY, int start, int stop);
```
```suggestion
void ArduinoGraphics::arc(int x, int y, int radiusX, int radiusY, int start, int stop)
```
Incorrect function syntax causes compilation to fail.

per1234 reviewed (changes requested):
Please always thoroughly test your code before submitting a pull request.

This sketch:
```c++
#include <ArduinoGraphics.h>

const byte canvasWidth = 12;
const byte canvasHeight = 12;

class ASCIIDrawClass : public ArduinoGraphics {
  public:
    ASCIIDrawClass() : ArduinoGraphics(canvasWidth, canvasHeight) {}

    virtual void set(int x, int y, uint8_t r, uint8_t g, uint8_t b) {
        if (x < 0 || x >= canvasWidth || y < 0 || y >= canvasHeight) {
    return;
  }
      // the r parameter is (mis)used to set the character to draw with
      _canvasBuffer[x][y] = r;
      (void)g;
      (void)b;
    }

    // display the drawing
    void endDraw() {
      ArduinoGraphics::endDraw();

      for (byte row = 0; row < canvasHeight; row++) {
        for (byte column = 0; column < canvasWidth; column++) {
          // handle unset parts of buffer
          if (_canvasBuffer[column][row] == 0) {
            _canvasBuffer[column][row] = ' ';
          }
          Serial.print(_canvasBuffer[column][row]);
        }
        Serial.println();
      }
    }

  private:
    char _canvasBuffer[canvasWidth][canvasHeight] = {0};
};

ASCIIDrawClass ASCIIDraw;

void setup() {
  Serial.begin(9600);
  while (!Serial);
  Serial.println("ellipse");
  ASCIIDraw.beginDraw();
  ASCIIDraw.background('X', 0, 0);
  ASCIIDraw.clear();
  ASCIIDraw.stroke('*', 0, 0);
  ASCIIDraw.fill('O', 0, 0);
  ASCIIDraw.ellipse(6, 6, 10, 10);
  ASCIIDraw.endDraw();
  Serial.println("circle");
  ASCIIDraw.beginDraw();
  ASCIIDraw.background('X', 0, 0);
  ASCIIDraw.clear();
  ASCIIDraw.stroke('*', 0, 0);
  ASCIIDraw.fill('O', 0, 0);
  ASCIIDraw.circle(6, 6, 5);
  ASCIIDraw.endDraw();
  Serial.println("arc");
  ASCIIDraw.beginDraw();
  ASCIIDraw.background('X', 0, 0);
  ASCIIDraw.clear();
  ASCIIDraw.stroke('*', 0, 0);
  ASCIIDraw.fill('O', 0, 0);
  ASCIIDraw.arc(6, 6, 10, 10, 0, 180);
  ASCIIDraw.endDraw();
}

void loop() {}
```
produces this output:
```
ellipse
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXX***XXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
circle
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
arc
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
XXXXXXXXXXXX
```
So there are definitely some serious problems with your code.

ashutoshshrm529 committed:
Changed and fixed as suggested

Co-Authored-By: per1234 <accounts@perglass.com>

Changed ellipse function. Removed arc function for now.

ashutoshshrm529 commented:
@per1234 , I apologise for the mistake in the previous pull, I have changed the ellipse function quite a bit and removed the arc function for now(i ran into a few issues with it). Please review this new commit

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino-libraries/ArduinoGraphics?pullRequest=11) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino-libraries/ArduinoGraphics?pullRequest=11) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino-libraries/ArduinoGraphics?pullRequest=11) it.</sub>

aentinger commented:
Closing this PR due to inactivity.
